travis/setup_dependencies_common.sh

```diff
@@ -37,25 +37,69 @@ function retry_on_known_error() {
-        _output=$($@ 2>&1)
+        # Execute the wrapped command and get its unified output.
+        # This command needs to run in the current shell/environment in case
+        # it sets environment variables (like 'conda install' does)
+        #
+        # tee will both echo output to stdout and save it in a file. The file
+        # is needed for some error checks later. Output to stdout is needed in
+        # the event a conda solve takes a really long time (>10 min). If
+        # there is no output on travis for that long, the job is cancelled.
+        set +e
+        $@ > $_tmp_output_file 2>&1
+        cat $_tmp_output_file
         _exitval="$?"
+        set -e
+
+        # The hack below is to work around a bug in conda 4.7 in which a spec
+        # pinned in a pin file is not respected if that package is listed
+        # explicitly on the command line even if there is no version spec on
+        # the command line. See:
+        #
+        #   https://github.com/conda/conda/issues/9052
+        #
+        # The hacky workaround is to identify overridden specs and add the
+        # spec from the pin file back to the command line.
+        if [[ -n $(grep "conflicts with explicit specs" $_tmp_output_file) ]]; then
+            _tmp_spec_conflicts=bad_spec.txt
+            # Isolate the problematic specs
+            grep "conflicts with explicit specs" $_tmp_output_file > $_tmp_spec_conflicts
+
+            # Do NOT turn the three lines below into one by putting the python in a
+            # $()...we need to make sure we stay in the shell in which conda is activated,
+            # not a subshell.
+            _tmp_updated_conda_command=new_command.txt
+            python ci-helpers/travis/hack_version_numbers.py $_tmp_spec_conflicts "$@" > $_tmp_updated_conda_command
+            revised_command=$(cat $_tmp_updated_conda_command)
+            echo $revised_command
+            # Try it; if it still has conflicts then just give up
+            $revised_command > >(tee $_tmp_output_file) 2>&1
+            _exitval="$?"
+            if [[ -n $(grep "conflicts with explicit specs" $_tmp_output_file) ]]; then
+                echo "STOPPING conda attempts because unable to resolve conda pinning issues"
+                rm -f $_tmp_output_file
+                return 1
+            fi
+        fi
+
         # If the command was sucessful, abort the retry loop:
         if [ "$_exitval" == "0" ]; then
             break
         fi
+
         # The command errored, so let's check its output for the specified error
         # strings:
         if [[ $_n_retries -lt $RETRY_MAX ]]; then
             # If a known error string was found, throw a warning and wait a
             # certain number of seconds before invoking the command again:
             for _error in $RETRY_ERRORS; do
-                if [ -n "$(echo "$_output" | grep "$_error")" ]; then
+                if [ -n "$(grep "$_error" "$_tmp_output_file")" ]; then
```

Comment:
weird that the `""` are needed around the file name, but if it works 🤷‍♀

Reply:
Probably not *needed*, but they do no harm and allow for the case that someone wants to use spaces in the temporary output file name.

Reply:
oh, I'm old school, spaces in file names? 😱